File: examples/python/basic_usage.py
"""
Basic usage examples for the Backstage Python SDK
"""

import asyncio
import time
from backstage_idp_sdk import create_client, ClientConfig

# Initialize client with configuration
config = ClientConfig(
    base_url='https://portal.company.com/api',
    api_key='your-api-key-here',
    timeout=30.0,
    retries=3
)

async def check_system_health():
    """Example 1: Check system health"""
    print("🏥 Checking system health...")
    
    async with create_client(config) as client:
        try:
            health = await client.system.get_health(verbose=True)
            print(f"System Status: {health.status}")
            print("Services:")
            for service_name, service_health in health.services.items():
                print(f"  - {service_name}: {service_health.status} ({service_health.message})")
            
            hours = health.uptime // 3600
            print(f"Uptime: {hours} hours")
        except Exception as error:
            print(f"Health check failed: {error}")

async def manage_plugins():
    """Example 2: List and install plugins"""
    print("🔌 Managing plugins...")
    
    async with create_client(config) as client:
        try:
            # List available plugins
            from backstage_idp_sdk.types import QueryParams
            
            params = QueryParams(category='catalog', limit=10)
            plugins = await client.plugins.list(params)
            
            print(f"Found {plugins.total} plugins:")
            for plugin in plugins.items:
                print(f"- {plugin.name} ({plugin.version}) - {plugin.status}")
            
            # Find and install a plugin
            catalog_plugin = next(
                (p for p in plugins.items if 'catalog' in p.name.lower()),
                None
            )
            
            if catalog_plugin and catalog_plugin.status == 'available':
                print(f"Installing {catalog_plugin.name}...")
                
                from backstage_idp_sdk.types import PluginInstallRequest
                
                install_request = PluginInstallRequest(
                    plugin_id=catalog_plugin.id,
                    version=catalog_plugin.version,
                    config={
                        'enabled': True,
                        'refreshInterval': '5m'
                    }
                )
                
                result = await client.plugins.install(install_request)
                print(f"Installation result: {result.success}")
                
        except Exception as error:
            print(f"Plugin management failed: {error}")

async def workflow_management():
    """Example 3: Create and execute workflows"""
    print("⚡ Managing workflows...")
    
    async with create_client(config) as client:
        try:
            from backstage_idp_sdk.types import WorkflowRequest, WorkflowStep
            
            # Create workflow steps
            build_step = WorkflowStep(
                id='build',
                name='Build Docker Image',
                type='docker-build',
                status='pending',
                config={
                    'dockerfile': 'Dockerfile',
                    'context': '.',
                    'tags': ['latest']
                }
            )
            
            deploy_step = WorkflowStep(
                id='deploy',
                name='Deploy to Kubernetes',
                type='k8s-deploy',
                status='pending',
                config={
                    'namespace': 'production',
                    'replicas': 3,
                    'healthCheck': '/health'
                }
            )
            
            # Create workflow
            workflow_request = WorkflowRequest(
                name='Deploy Service',
                description='Automated service deployment workflow',
                steps=[build_step, deploy_step]
            )
            
            workflow = await client.workflows.create(workflow_request)
            print(f"Created workflow: {workflow.id}")
            
            # Execute workflow
            from backstage_idp_sdk.types import WorkflowExecutionRequest
            
            execution_request = WorkflowExecutionRequest(
                parameters={
                    'environment': 'staging',
                    'version': '1.2.3'
                }
            )
            
            execution = await client.workflows.execute(workflow.id, execution_request)
            print(f"Workflow execution started: {execution.execution_id}")
            
            # Monitor execution status
            status = execution.status
            while status in ['pending', 'running']:
                await asyncio.sleep(5)
                
                updated_workflow = await client.workflows.get(workflow.id)
                status = updated_workflow.status
                print(f"Workflow status: {status}")
                
        except Exception as error:
            print(f"Workflow management failed: {error}")

async def search_resources():
    """Example 4: Search portal resources"""
    print("🔍 Searching resources...")
    
    async with create_client(config) as client:
        try:
            search_results = await client.search('microservice', 'service')
            
            print(f"Found {search_results.total} results:")
            for result in search_results.results:
                print(f"- {result.title} ({result.type}) - Score: {result.score}")
                if result.description:
                    print(f"  {result.description}")
                    
        except Exception as error:
            print(f"Search failed: {error}")

async def handle_notifications():
    """Example 5: Manage notifications"""
    print("🔔 Managing notifications...")
    
    async with create_client(config) as client:
        try:
            # Get recent notifications
            notifications_response = await client.notifications.list(
                unread_only=False, 
                limit=10
            )
            
            notifications = notifications_response['notifications']
            unread_count = notifications_response.get('unreadCount', 0)
            
            print(f"You have {unread_count} unread notifications:")
            for notification in notifications:
                status = '✓' if notification.read else '○'
                print(f"{status} {notification.title} - {notification.type}")
                print(f"  {notification.message}")
            
            # Mark first notification as read
            if notifications and not notifications[0].read:
                await client.notifications.mark_as_read(notifications[0].id)
                print("Marked notification as read")
            
            # Create a new notification
            from backstage_idp_sdk.types import NotificationRequest
            
            notification_request = NotificationRequest(
                title='Python SDK Example',
                message='This is a test notification from the Python SDK',
                type='info'
            )
            
            await client.notifications.create(notification_request)
            print("Created test notification")
            
        except Exception as error:
            print(f"Notification management failed: {error}")

async def manage_tenants():
    """Example 6: Tenant management"""
    print("🏢 Managing tenants...")
    
    async with create_client(config) as client:
        try:
            # Get current tenant info
            current_tenant_response = await client.tenants.get_current()
            tenant = current_tenant_response['tenant']
            user_role = current_tenant_response.get('userRole')
            is_owner = current_tenant_response.get('isOwner', False)
            
            print(f"Current tenant: {tenant.name}")
            print(f"User role: {user_role}")
            print(f"Is owner: {is_owner}")
            
            # Get tenant analytics
            analytics_response = await client.tenants.get_analytics(30)
            analytics = analytics_response['analytics']
            
            print("Tenant analytics (30 days):")
            print(f"- Users: {analytics.user_count}")
            print(f"- Plugins: {analytics.plugin_count}")
            print(f"- API requests: {analytics.api_requests}")
            print(f"- Storage used: {analytics.storage_used} MB")
            
        except Exception as error:
            print(f"Tenant management failed: {error}")

async def use_graphql_queries():
    """Example 7: GraphQL queries"""
    print("📊 Using GraphQL queries...")
    
    async with create_client(config) as client:
        try:
            # Use GraphQL client for complex queries
            plugins_query = """
                query GetPlugins($limit: Int, $category: String) {
                    plugins(limit: $limit, category: $category) {
                        items {
                            id
                            name
                            version
                            status
                            description
                        }
                        total
                    }
                }
            """
            
            plugins_result = await client.graphql.query(
                plugins_query,
                variables={'limit': 5, 'category': 'monitoring'}
            )
            
            print("GraphQL plugins result:")
            if 'plugins' in plugins_result:
                for plugin in plugins_result['plugins']['items']:
                    print(f"- {plugin['name']} ({plugin['version']})")
            
            # Health query
            health_query = """
                query GetSystemHealth($verbose: Boolean) {
                    systemHealth(verbose: $verbose) {
                        status
                        timestamp
                        version
                        uptime
                        services {
                            database { status message }
                            backstage { status message }
                            cache { status message }
                        }
                    }
                }
            """
            
            health_result = await client.graphql.query(
                health_query,
                variables={'verbose': True}
            )
            
            if 'systemHealth' in health_result:
                health = health_result['systemHealth']
                print(f"GraphQL health result: {health['status']}")
            
        except Exception as error:
            print(f"GraphQL operations failed: {error}")

async def batch_operations():
    """Example 8: Batch operations"""
    print("📦 Running batch operations...")
    
    async with create_client(config) as client:
        try:
            # Execute multiple operations in batch
            operations = [
                {
                    'method': 'GET',
                    'path': '/health',
                    'data': None
                },
                {
                    'method': 'GET', 
                    'path': '/plugins',
                    'data': None
                },
                {
                    'method': 'GET',
                    'path': '/workflows',
                    'data': None
                }
            ]
            
            results = await client.batch_execute(operations)
            
            print("Batch operation results:")
            for i, result in enumerate(results):
                print(f"Operation {i+1}: Status {result.get('status', 'unknown')}")
                
        except Exception as error:
            print(f"Batch operations failed: {error}")

async def handle_realtime_events():
    """Example 9: Real-time event handling"""
    print("📡 Setting up real-time event handling...")
    
    async with create_client(config) as client:
        try:
            # Connect WebSocket
            await client.connect()
            
            # Subscribe to events
            def on_plugin_event(event):
                print(f"Plugin event: {event['type']} - {event.get('data', {})}")
            
            def on_workflow_event(event):
                print(f"Workflow event: {event['type']} - {event.get('data', {})}")
            
            def on_system_event(event):
                print(f"System event: {event['type']} - {event.get('data', {})}")
            
            # Set up event handlers
            plugin_sub = await client.events.on_plugin_event(on_plugin_event)
            workflow_sub = await client.events.on_workflow_event(on_workflow_event)
            system_sub = await client.events.on_system_event(on_system_event)
            
            print("Event subscriptions active. Listening for events...")
            
            # Listen for events for 30 seconds
            await asyncio.sleep(30)
            
            # Clean up subscriptions
            if plugin_sub:
                await client.events.unsubscribe(plugin_sub)
            if workflow_sub:
                await client.events.unsubscribe(workflow_sub)
            if system_sub:
                await client.events.unsubscribe(system_sub)
            
        except Exception as error:
            print(f"Real-time event handling failed: {error}")

async def main():
    """Main execution function"""
    print("🚀 Starting Backstage Python SDK examples...\n")
    
    examples = [
        ("System Health", check_system_health),
        ("Plugin Management", manage_plugins),
        ("Workflow Management", workflow_management),
        ("Search Resources", search_resources),
        ("Notifications", handle_notifications),
        ("Tenant Management", manage_tenants),
        ("GraphQL Queries", use_graphql_queries),
        ("Batch Operations", batch_operations),
    ]
    
    # Run all examples
    for name, example_func in examples:
        try:
            print(f"\n=== {name} ===")
            await example_func()
        except Exception as error:
            print(f"Example '{name}' failed: {error}")
        
        # Wait between examples
        await asyncio.sleep(1)
    
    # Run real-time example last (it has its own cleanup)
    print("\n=== Real-time Events ===")
    await handle_realtime_events()
    
    print("\n✅ All examples completed!")

if __name__ == "__main__":
    # Run the examples
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n⏹️ Examples interrupted by user")
    except Exception as error:
        print(f"\n❌ Example execution failed: {error}")
        exit(1)

File: examples/typescript/basic-usage.ts
/**
 * Basic usage examples for the Backstage TypeScript SDK
 */

import { createBackstageClient } from '@backstage-idp/sdk-typescript';

// Initialize client with API key
const client = createBackstageClient({
  baseURL: 'https://portal.company.com/api',
  apiKey: 'your-api-key-here',
  timeout: 30000,
  retries: 3
});

// Example 1: Check system health
async function checkSystemHealth() {
  console.log('🏥 Checking system health...');
  
  try {
    const health = await client.system.getHealth(true);
    console.log('System Status:', health.status);
    console.log('Services:', health.services);
    console.log('Uptime:', `${Math.floor(health.uptime / 3600)} hours`);
  } catch (error) {
    console.error('Health check failed:', error.message);
  }
}

// Example 2: List and install plugins
async function managePlugins() {
  console.log('🔌 Managing plugins...');
  
  try {
    // List available plugins
    const plugins = await client.plugins.list({
      category: 'catalog',
      limit: 10
    });
    
    console.log(`Found ${plugins.total} plugins:`);
    plugins.items.forEach(plugin => {
      console.log(`- ${plugin.name} (${plugin.version}) - ${plugin.status}`);
    });
    
    // Install a plugin
    const catalogPlugin = plugins.items.find(p => p.name.includes('catalog'));
    if (catalogPlugin && catalogPlugin.status === 'available') {
      console.log(`Installing ${catalogPlugin.name}...`);
      
      const result = await client.plugins.install({
        pluginId: catalogPlugin.id,
        version: catalogPlugin.version,
        config: {
          enabled: true,
          refreshInterval: '5m'
        }
      });
      
      console.log('Installation result:', result);
    }
  } catch (error) {
    console.error('Plugin management failed:', error.message);
  }
}

// Example 3: Create and execute workflows
async function workflowManagement() {
  console.log('⚡ Managing workflows...');
  
  try {
    // Create a new workflow
    const workflow = await client.workflows.create({
      name: 'Deploy Service',
      description: 'Automated service deployment workflow',
      steps: [
        {
          id: 'build',
          name: 'Build Docker Image',
          type: 'docker-build',
          status: 'pending',
          config: {
            dockerfile: 'Dockerfile',
            context: '.',
            tags: ['latest']
          }
        },
        {
          id: 'deploy',
          name: 'Deploy to Kubernetes',
          type: 'k8s-deploy',
          status: 'pending',
          config: {
            namespace: 'production',
            replicas: 3,
            healthCheck: '/health'
          }
        }
      ]
    });
    
    console.log('Created workflow:', workflow.id);
    
    // Execute the workflow
    const execution = await client.workflows.execute(workflow.id, {
      parameters: {
        environment: 'staging',
        version: '1.2.3'
      }
    });
    
    console.log('Workflow execution started:', execution.executionId);
    
    // Monitor execution status
    let status = execution.status;
    while (status === 'pending' || status === 'running') {
      await new Promise(resolve => setTimeout(resolve, 5000));
      
      const updated = await client.workflows.get(workflow.id);
      status = updated.status;
      console.log('Workflow status:', status);
    }
    
  } catch (error) {
    console.error('Workflow management failed:', error.message);
  }
}

// Example 4: Search portal resources
async function searchResources() {
  console.log('🔍 Searching resources...');
  
  try {
    const searchResults = await client.search('microservice', 'service');
    
    console.log(`Found ${searchResults.total} results:`);
    searchResults.results.forEach(result => {
      console.log(`- ${result.title} (${result.type}) - Score: ${result.score}`);
      if (result.description) {
        console.log(`  ${result.description}`);
      }
    });
  } catch (error) {
    console.error('Search failed:', error.message);
  }
}

// Example 5: Manage notifications
async function handleNotifications() {
  console.log('🔔 Managing notifications...');
  
  try {
    // Get recent notifications
    const notifications = await client.notifications.list(false, 10);
    
    console.log(`You have ${notifications.unreadCount} unread notifications:`);
    notifications.notifications.forEach(notification => {
      const status = notification.read ? '✓' : '○';
      console.log(`${status} ${notification.title} - ${notification.type}`);
      console.log(`  ${notification.message}`);
    });
    
    // Mark first notification as read
    if (notifications.notifications.length > 0 && !notifications.notifications[0].read) {
      await client.notifications.markAsRead(notifications.notifications[0].id);
      console.log('Marked notification as read');
    }
    
    // Create a new notification
    await client.notifications.create({
      title: 'SDK Example',
      message: 'This is a test notification from the TypeScript SDK',
      type: 'info'
    });
    
    console.log('Created test notification');
    
  } catch (error) {
    console.error('Notification management failed:', error.message);
  }
}

// Example 6: Real-time event handling with WebSocket
async function handleRealtimeEvents() {
  console.log('📡 Setting up real-time event handling...');
  
  try {
    // Connect to WebSocket
    await client.connect();
    
    // Subscribe to plugin events
    const pluginSubscription = client.events.onPluginEvent((event) => {
      console.log('Plugin event:', event.type, event.data);
    });
    
    // Subscribe to workflow events
    const workflowSubscription = client.events.onWorkflowEvent((event) => {
      console.log('Workflow event:', event.type, event.data);
    });
    
    // Subscribe to system events
    const systemSubscription = client.events.onSystemEvent((event) => {
      console.log('System event:', event.type, event.data);
    });
    
    console.log('Event subscriptions active. Listening for events...');
    
    // Keep alive for 30 seconds to receive events
    setTimeout(() => {
      console.log('Cleaning up subscriptions...');
      client.disconnect();
    }, 30000);
    
  } catch (error) {
    console.error('Real-time event handling failed:', error.message);
  }
}

// Example 7: Tenant management
async function manageTenants() {
  console.log('🏢 Managing tenants...');
  
  try {
    // Get current tenant info
    const currentTenant = await client.tenants.getCurrent();
    console.log('Current tenant:', currentTenant.tenant.name);
    console.log('User role:', currentTenant.userRole);
    console.log('Is owner:', currentTenant.isOwner);
    
    // Get tenant analytics
    const analytics = await client.tenants.getAnalytics(30);
    console.log('Tenant analytics (30 days):');
    console.log('- Users:', analytics.analytics.userCount);
    console.log('- Plugins:', analytics.analytics.pluginCount);
    console.log('- API requests:', analytics.analytics.apiRequests);
    console.log('- Storage used:', `${analytics.analytics.storageUsed} MB`);
    
  } catch (error) {
    console.error('Tenant management failed:', error.message);
  }
}

// Example 8: GraphQL queries
async function useGraphQLQueries() {
  console.log('📊 Using GraphQL queries...');
  
  try {
    // Use pre-defined query for plugins
    const pluginsData = await client.graphql.query(
      client.graphql.queries.getPlugins,
      { limit: 5, category: 'monitoring' }
    );
    
    console.log('GraphQL plugins result:', pluginsData.plugins);
    
    // Use pre-defined query for system health
    const healthData = await client.graphql.query(
      client.graphql.queries.getSystemHealth,
      { verbose: true }
    );
    
    console.log('GraphQL health result:', healthData.systemHealth);
    
    // Execute a custom GraphQL mutation
    const notificationResult = await client.graphql.mutate(
      client.graphql.mutations.createNotification,
      {
        title: 'GraphQL Test',
        message: 'This notification was created using GraphQL',
        type: 'INFO'
      }
    );
    
    console.log('GraphQL mutation result:', notificationResult);
    
  } catch (error) {
    console.error('GraphQL operations failed:', error.message);
  }
}

// Main execution function
async function main() {
  console.log('🚀 Starting Backstage SDK examples...\n');
  
  const examples = [
    { name: 'System Health', fn: checkSystemHealth },
    { name: 'Plugin Management', fn: managePlugins },
    { name: 'Workflow Management', fn: workflowManagement },
    { name: 'Search Resources', fn: searchResources },
    { name: 'Notifications', fn: handleNotifications },
    { name: 'Tenant Management', fn: manageTenants },
    { name: 'GraphQL Queries', fn: useGraphQLQueries }
  ];
  
  // Run all examples
  for (const example of examples) {
    try {
      console.log(`\n=== ${example.name} ===`);
      await example.fn();
    } catch (error) {
      console.error(`Example "${example.name}" failed:`, error.message);
    }
    
    // Wait between examples
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  
  // Run real-time example last (it has its own cleanup)
  console.log('\n=== Real-time Events ===');
  await handleRealtimeEvents();
  
  console.log('\n✅ All examples completed!');
}

// Error handling wrapper
async function runExamples() {
  try {
    await main();
  } catch (error) {
    console.error('❌ Example execution failed:', error);
    process.exit(1);
  } finally {
    // Cleanup
    client.dispose();
  }
}

// Run if this file is executed directly
if (require.main === module) {
  runExamples();
}

export {
  checkSystemHealth,
  managePlugins,
  workflowManagement,
  searchResources,
  handleNotifications,
  handleRealtimeEvents,
  manageTenants,
  useGraphQLQueries
};